pkg/manager/manager.go

```diff
@@ -0,0 +1,167 @@
+}
+
+// Options are the arguments for creating a new Manager
+type Options struct {
+	// Scheme is the scheme used to resolve runtime.Objects to GroupVersionKinds / Resources
+	// Defaults to the kubernetes/client-go scheme.Scheme
+	Scheme *runtime.Scheme
+
+	// Mapper is the rest mapper used to map go types to Kubernetes APIs
+	MapperProvider func(c *rest.Config) (meta.RESTMapper, error)
+
+	// Dependency injection for testing
+	newCache            func(config *rest.Config, opts cache.Options) (cache.Cache, error)
+	newClient           func(config *rest.Config, options client.Options) (client.Client, error)
+	newRecorderProvider func(config *rest.Config, scheme *runtime.Scheme) (recorder.Provider, error)
+}
+
+// Runnable allows a component to be started.
+type Runnable interface {
+	// Start starts running the component.  The component will stop running when the channel is closed.
+	// Start blocks until the channel is closed or an error occurs.
+	Start(<-chan struct{}) error
+}
+
+// RunnableFunc implements Runnable
+type RunnableFunc func(<-chan struct{}) error
+
+// Start implements Runnable
+func (r RunnableFunc) Start(s <-chan struct{}) error {
+	return r(s)
+}
+
+// New returns a new Manager
+func New(config *rest.Config, options Options) (Manager, error) {
+	cm := &controllerManager{config: config, scheme: options.Scheme, errChan: make(chan error)}
+
+	// Initialize a rest.config if none was specified
+	if cm.config == nil {
+		return nil, fmt.Errorf("must specify Config")
+	}
+
+	// Use the Kubernetes client-go scheme if none is specified
+	if cm.scheme == nil {
+		cm.scheme = scheme.Scheme
+	}
+
+	// Set default values for options fields
+	options = setOptionsDefaults(options)
+
+	mapper, err := options.MapperProvider(cm.config)
+	if err != nil {
+		log.Error(err, "Failed to get API Group-Resources")
+		return nil, err
+	}
+
+	// Create the Client for Write operations.
+	writeObj, err := options.newClient(cm.config, client.Options{Scheme: cm.scheme, Mapper: mapper})
+	if err != nil {
+		return nil, err
+	}
+
+	cm.cache, err = options.newCache(cm.config, cache.Options{Scheme: cm.scheme, Mapper: mapper})
+	if err != nil {
+		return nil, err
+	}
+
+	cm.fieldIndexes = cm.cache
+	cm.client = client.DelegatingClient{Reader: cm.cache, Writer: writeObj}
+
+	// Create the recorder provider to inject event recorders for the components.
+	cm.recorderProvider, err = options.newRecorderProvider(cm.config, cm.scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
+// setOptionsDefaults set default values for Options fields
+func setOptionsDefaults(options Options) Options {
+	if options.MapperProvider == nil {
+		options.MapperProvider = apiutil.NewDiscoveryRESTMapper
+	}
+
+	// Allow newClient to be mocked
+	if options.newClient == nil {
+		options.newClient = client.New
+	}
+
+	// Allow newCache to be mocked
+	// TODO(directxman12): Figure out how to allow users to request a client without requesting a watch
```

Comment:
Don't know this TODO is still valid ? I remember some discussion where it came as *already* done ?

Reply:
Removed